three or more variable paired scale
Statistical test type
- paired : compare between variables (before and after), (compare multiple variables)
- unpaired: relationship between variables

In [ ]:
import numpy
import pandas
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris

In [ ]:
X, y = load_iris(return_X_y=True, as_frame=True)

In [ ]:
X.head(n=10)

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2
5,5.4,3.9,1.7,0.4
6,4.6,3.4,1.4,0.3
7,5.0,3.4,1.5,0.2
8,4.4,2.9,1.4,0.2
9,4.9,3.1,1.5,0.1


In [ ]:
y.head(n=3)


,target
0,0
1,0
2,0


In [ ]:
data = pandas.concat([X, y], axis=1)
data

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0
...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,2
146,6.3,2.5,5.0,1.9,2
147,6.5,3.0,5.2,2.0,2
148,6.2,3.4,5.4,2.3,2


In [ ]:
data['target'] = data['target'].replace({0:'Setosa', 1:'Versicolor', 2: 'Virginica'})

In [ ]:
data.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'target'],
      dtype='object')

In [ ]:
data.groupby('target').agg({
    'sepal length (cm)': ['mean', 'std'],
    'sepal width (cm)': ['mean', 'std'],
    'petal length (cm)': ['mean', 'std'],
    'petal width (cm)': ['mean', 'std']
})

sepal length (cm)           sepal width (cm)            \
                        mean       std             mean       std   
target                                                              
Setosa                 5.006  0.352490            3.428  0.379064   
Versicolor             5.936  0.516171            2.770  0.313798   
Virginica              6.588  0.635880            2.974  0.322497   

           petal length (cm)           petal width (cm)            
                        mean       std             mean       std  
target                                                             
Setosa                 1.462  0.173664            0.246  0.105386  
Versicolor             4.260  0.469911            1.326  0.197753  
Virginica              5.552  0.551895            2.026  0.274650

#Multivariate analysis of variance MANOVA

In [ ]:
from statsmodels.multivariate.manova import MANOVA

In [ ]:
data.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'target'],
      dtype='object')

In [ ]:
data = data.rename(columns={
    'sepal length (cm)':'sepal_length', 'sepal width (cm)':'sepal_width', 'petal length (cm)':'petal_length',
       'petal width (cm)':'petal_width'
})
data

,sepal_length,sepal_width,petal_length,petal_width,target
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa
...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,Virginica
146,6.3,2.5,5.0,1.9,Virginica
147,6.5,3.0,5.2,2.0,Virginica
148,6.2,3.4,5.4,2.3,Virginica


In [ ]:
data.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'target'], dtype='object')

In [ ]:
result = MANOVA.from_formula('sepal_length + sepal_width + petal_length + petal_width ~ target', data=data)
print(result.mv_test())

                   Multivariate linear model
                                                                
----------------------------------------------------------------
       Intercept         Value  Num DF  Den DF   F Value  Pr > F
----------------------------------------------------------------
          Wilks' lambda  0.0170 4.0000 144.0000 2086.7720 0.0000
         Pillai's trace  0.9830 4.0000 144.0000 2086.7720 0.0000
 Hotelling-Lawley trace 57.9659 4.0000 144.0000 2086.7720 0.0000
    Roy's greatest root 57.9659 4.0000 144.0000 2086.7720 0.0000
----------------------------------------------------------------
                                                                
----------------------------------------------------------------
         target          Value  Num DF  Den DF   F Value  Pr > F
----------------------------------------------------------------
          Wilks' lambda  0.0234 8.0000 288.0000  199.1453 0.0000
         Pillai's trace  1.1919 8.0000 290.00

'''
H0 - no difference in sepal_length,sepal_width,petal_length, petal_width for flower groups-setosa, virginica,vertose

Ha - there is difference in sepal_length,sepal_width,petal_length, petal_width for flowers-setosa, virginica,vertose
'''

pvalue = 0.000
'''
         target          Value  Num DF  Den DF   F Value  Pr > F
----------------------------------------------------------------
          Wilks' lambda  0.0234 8.0000 288.0000  199.1453 0.0000

null hypothesis rejected
'''

In [ ]:
wilks_lambda = 0.0234

if wilks_lambda > 0.9:
    print("Effect Size: Small")
elif wilks_lambda > 0.7:
    print("Effect Size: Medium")
else:
    print("Effect Size: Large")

Effect Size: Large
